### PR TITLE
Add advanced rendering API for modders

### PR DESCRIPTION
**Pull request by eddoww**
## Summary

- Implements comprehensive rendering features for native mod development (DLL_EXPORT API)
- Adds alpha blending primitives, shape primitives, curves, gradients, blend modes, custom texture loading, screen effects, and render targets
- Security hardened with path traversal protection, size limits, and overflow prevention
- Includes automated test suite for all new primitives

### Basic Drawing Primitives
- `render_pixel_alpha()` - Draw single pixel with alpha
- `render_line_alpha()` - Draw line with alpha
- `render_line_aa()` - Anti-aliased line (Xiaolin Wu algorithm)
- `render_thick_line_alpha()` - Variable thickness line

### Rectangle Primitives
- `render_rect_alpha()` - Draw filled rectangle with alpha
- `render_rect_outline_alpha()` - Rectangle outline only
- `render_rounded_rect_alpha()` - Rounded rectangle outline
- `render_rounded_rect_filled_alpha()` - Filled rounded rectangle

### Circle and Ellipse Primitives
- `render_circle_alpha()` - Circle outline
- `render_circle_filled_alpha()` - Filled circle
- `render_ellipse_alpha()` - Ellipse outline
- `render_ellipse_filled_alpha()` - Filled ellipse
- `render_ring_alpha()` - Arc with inner/outer radius (annulus)

### Triangle Primitives
- `render_triangle_alpha()` - Triangle outline
- `render_triangle_filled_alpha()` - Filled triangle (scanline algorithm)

### Arc and Curve Primitives
- `render_arc_alpha()` - Circular arc segment
- `render_bezier_quadratic_alpha()` - Quadratic bezier curve (3 control points)
- `render_bezier_cubic_alpha()` - Cubic bezier curve (4 control points)

### Gradient Primitives
- `render_gradient_rect_h()` - Horizontal color gradient rectangle
- `render_gradient_rect_v()` - Vertical color gradient rectangle
- `render_gradient_circle()` - Radial gradient (glow/bloom effects)

### Blend Modes
- `render_set_blend_mode()` / `render_get_blend_mode()`
- Supports: 
- BLEND_NORMAL (0): Alpha blending (default)
- BLEND_ADDITIVE (1): Additive blending for glow effects
- BLEND_MOD (2): Color modulate for tinting
- BLEND_MUL (3): Color multiply with alpha
- BLEND_NONE (4): No blending, direct overwrite

### Custom Texture Loading
- `render_load_texture()` - Load PNG from file (with path security)
- `render_unload_texture()` - Free texture memory
- `render_texture()` / `render_texture_scaled()` - Render texture
- `render_texture_width()` / `render_texture_height()` - Query dimensions

### Screen Effects
- `render_screen_tint()` - Color overlay effect
- `render_vignette()` - Cinematic edge darkening (optimized)
- `render_screen_flash()` - Damage/healing visual feedback

### Render Targets (Offscreen Rendering)
- `render_create_target()` - Create offscreen buffer (size-limited)
- `render_destroy_target()` - Free render target
- `render_set_target()` - Redirect rendering to target
- `render_target_to_screen()` - Blit target to screen
- `render_clear_target()` - Clear to transparent

## Security Measures

- **Path traversal protection**: Rejects `../`, absolute paths, null bytes
- **Render target size limits**: Max 4096x4096 to prevent memory exhaustion
- **Scale validation**: Max 100x to prevent integer overflow

## Test Plan

- [x] All new primitives tested via `test_render_primitives.c`
- [x] Edge cases covered (zero radius, degenerate shapes, alpha extremes)
- [x] Stress test with many draw calls
- [x] Build passes with strict warnings (`-Werror`)
- [x] All existing tests pass (`make test`)

### CODE REVIEW

**Comment by eddoww** [comment]
Now rebased on SDL3, and converted to SDL3 calls. Would be great if we can get a release going with these changes (SDL3 and these modding API).

**Review by DanielBrockhaus — APPROVED** [review]

File: src/sdl/sdl_draw.c [hunk 1]
```diff
@@ -459,69 +463,1658 @@ void sdl_bargraph(int sx, int sy, int dx, unsigned char *data, int x_offset, int
...
+}
+
+void sdl_set_blend_mode(int mode)
+{
+	switch (mode) {
+	case 0: // BLEND_NORMAL
+		current_blend_mode = SDL_BLENDMODE_BLEND;
+		break;
+	case 1: // BLEND_ADDITIVE
+		current_blend_mode = SDL_BLENDMODE_ADD;
```
Comment (PrismaPhonic) [inline src/sdl/sdl_draw.c]: If a mod sets `BLEND_ADDITIVE` and crashes (or forgets to reset), all subsequent rendering uses additive blending until client restart.

**Suggested Fix:**

**Option A:** Auto-reset per frame 
```c
// In sdl_core.c or render.c, at frame end:
void render_frame_end(void) {
    sdl_set_blend_mode(BLEND_NORMAL);  // Reset to default
}
```

**Option B:** Save/restore API (more flexible)
```c
// Stack-based state management
static SDL_BlendMode blend_stack[8];
static int blend_stack_top = -1;

DLL_EXPORT void render_push_blend_mode(int mode) {
    if (blend_stack_top < 7) {
        blend_stack[++blend_stack_top] = current_blend_mode;
        render_set_blend_mode(mode);
    }
}

DLL_EXPORT void render_pop_blend_mode(void) {
    if (blend_stack_top >= 0) {
        current_blend_mode = blend_stack[blend_stack_top--];
        SDL_SetRenderDrawBlendMode(sdlren, current_blend_mode);
    }
}
```

Then mods can:
```c
render_push_blend_mode(BLEND_ADDITIVE);
// ... draw glowy stuff ...
render_pop_blend_mode();  // Auto-restore
```

File: src/sdl/sdl_draw.c [hunk 2]
```diff
@@ -459,69 +463,1658 @@ void sdl_bargraph(int sx, int sy, int dx, unsigned char *data, int x_offset, int
...
+	if (strlen(path) >= 2 && path[1] == ':') {
+		warn("mod texture: absolute paths not allowed: %s", path);
+		return 0;
+	}
+
+	// Reject path traversal sequences
+	p = path;
+	while (*p) {
+		// Check for ".." at start or after path separator
+		if (p[0] == '.' && p[1] == '.') {
```
Comment (PrismaPhonic) [inline src/sdl/sdl_draw.c]: Mods might still be able to read arbitrary files.

The check only rejects `..` when it's a complete path component (starts at beginning or after `/` or `\`, and ends at end or before `/` or `\`).

Potential bypasses:
```c
render_load_texture("..././etc/passwd");      // p[2]='.' not '/' so not rejected
render_load_texture("....//etc/passwd");      // p[2]='.' not '/' so not rejected
```

Whether these actually bypass depends on how the filesystem and `sdl_load_png` handle these paths. The validation code doesn't catch them, but the OS might normalize them away. Needs testing to confirm if this is exploitable.

If you want to be defensive here:

```c
#include <limits.h>  // For PATH_MAX
#include <stdlib.h>  // For realpath

static int validate_mod_path(const char *path) {
    char canonical[PATH_MAX];
    const char *allowed_base = "/path/to/game/mods";  // Configure this
    
    // Get canonical absolute path (resolves .., symlinks, etc)
    if (!realpath(path, canonical)) {
        warn("mod texture: invalid path: %s", path);
        return 0;
    }
    
    // Check if it's inside allowed directory
    size_t base_len = strlen(allowed_base);
    if (strncmp(canonical, allowed_base, base_len) != 0) {
        warn("mod texture: path outside mods directory: %s", path);
        return 0;
    }
    
    return 1;
}
```

File: src/sdl/sdl_draw.c [hunk 3]
```diff
@@ -459,69 +463,1658 @@ void sdl_bargraph(int sx, int sy, int dx, unsigned char *data, int x_offset, int
...
+
+	float width = (float)(ex - sx);
+	for (int x = sx; x < ex; x++) {
+		float t = (float)(x - sx) / width;
+		int r = (int)(r1 + t * (r2 - r1));
+		int g = (int)(g1 + t * (g2 - g1));
+		int b = (int)(b1 + t * (b2 - b1));
+
+		SDL_SetRenderDrawColor(sdlren, (Uint8)r, (Uint8)g, (Uint8)b, (Uint8)alpha);
+		SDL_RenderLine(sdlren, (float)((x + x_offset) * sdl_scale), (float)((sy + y_offset) * sdl_scale),
```
Comment (PrismaPhonic) [inline src/sdl/sdl_draw.c]: This issues two SDL calls per column.
- 800px gradient = 1,600 SDL calls (800 * SetColor + 800 * RenderLine)
- 1920px gradient = 3,840 SDL calls
- Each call has function call overhead + driver overhead

**Faster approach using SDL_RenderGeometry:**

```c
void sdl_gradient_rect_h(int sx, int sy, int ex, int ey, unsigned short color1, unsigned short color2,
    unsigned char alpha, int clipsx, int clipsy, int clipex, int clipey, int x_offset, int y_offset)
{
    // Apply clipping
    if (sx < clipsx) sx = clipsx;
    if (sy < clipsy) sy = clipsy;
    if (ex > clipex) ex = clipex;
    if (ey > clipey) ey = clipey;
    if (sx >= ex || sy >= ey) return;

    SDL_SetRenderDrawBlendMode(sdlren, current_blend_mode);

    Uint8 r1 = (Uint8)R16TO32(color1), g1 = (Uint8)G16TO32(color1), b1 = (Uint8)B16TO32(color1);
    Uint8 r2 = (Uint8)R16TO32(color2), g2 = (Uint8)G16TO32(color2), b2 = (Uint8)B16TO32(color2);

    // Apply offset and scale
    float fsx = (float)((sx + x_offset) * sdl_scale);
    float fsy = (float)((sy + y_offset) * sdl_scale);
    float fex = (float)((ex + x_offset) * sdl_scale);
    float fey = (float)((ey - 1 + y_offset) * sdl_scale);

    // Create vertices for a single quad with vertex colors
    SDL_Vertex vertices[4] = {
        {{fsx, fsy}, {r1, g1, b1, alpha}, {0, 0}},  // Top-left
        {{fex, fsy}, {r2, g2, b2, alpha}, {0, 0}},  // Top-right
        {{fex, fey}, {r2, g2, b2, alpha}, {0, 0}},  // Bottom-right
        {{fsx, fey}, {r1, g1, b1, alpha}, {0, 0}}   // Bottom-left
    };

    Sint32 indices[6] = {0, 1, 2, 0, 2, 3};

    SDL_RenderGeometry(sdlren, NULL, vertices, 4, indices, 6);
}
```

**Benefits:**
- Single SDL call instead of 1000s
- GPU interpolates colors for you (smoother gradients)
- Same approach works for `sdl_gradient_rect_v` (just swap vertex colors)

File: src/game/render.c [hunk 4]
```diff
@@ -684,6 +720,606 @@ DLL_EXPORT void render_pixel(int x, int y, unsigned short col)
...
+{
+	int border_size, i, band_start, band_end, band_count;
+	unsigned char alpha;
+
+	if (intensity == 0) {
+		return;
+	}
+
+	// Calculate border size based on intensity (max ~15% of screen)
+	border_size = (XRES < YRES ? XRES : YRES) * intensity / 1700;
```
Comment (PrismaPhonic) [inline src/game/render.c]: Why 1700? A comment here would be useful

File: tests/test_render_primitives.c [hunk 5]
```diff
@@ -0,0 +1,447 @@
+/*
```
Comment (PrismaPhonic) [inline tests/test_render_primitives.c]: Your test file is excellent for crash/stability testing, but we should add a few more tests ideally:

**Tests to add:**
```c
// Test that blend mode doesn't persist across frames
TEST(test_blend_mode_frame_isolation) {
    sdl_set_blend_mode(BLEND_ADDITIVE);
    sdl_circle_filled_alpha(100, 100, 30, 0x7FFF, 128, 0, 0);
    
    // Simulate frame end/start
    render_frame_end();
    render_frame_start();
    
    // Blend mode should be reset to BLEND_NORMAL
    ASSERT_EQ_INT(BLEND_NORMAL, sdl_get_blend_mode());
}

// Test that circles scale correctly with sdl_scale
TEST(test_circle_scaling) {
    int old_scale = sdl_scale;
    
    // At scale=1, radius 50 should draw 50-pixel radius
    sdl_scale = 1;
    sdl_circle_alpha(100, 100, 50, 0x7FFF, 255, 0, 0);
    // (Would need to verify rightmost point is at x=150)
    
    // At scale=2, radius 50 should draw 100-pixel radius
    sdl_scale = 2;
    sdl_circle_alpha(100, 100, 50, 0x7FFF, 255, 0, 0);
    // (Would need to verify rightmost point is at x=300, not x=250)
    
    sdl_scale = old_scale;
}

// Test that line clipping preserves slope
TEST(test_line_clipping_slope) {
    // Diagonal line from outside clip rect should maintain 45 deg angle when clipped
    sdl_line_alpha(-100, -100, 900, 900, 0x7FFF, 255, 0, 0, 800, 600, 0, 0);
    
    // Expected: Line exits at (600, 600) maintaining slope 1.0
    // Current bug: Line exits at (799, 599) with slope ~0.75
}

// Test path validation security
TEST(test_mod_texture_path_validation) {
    // These should be rejected
    ASSERT_EQ_INT(-1, sdl_load_mod_texture("/etc/passwd"));
    ASSERT_EQ_INT(-1, sdl_load_mod_texture("../../../etc/passwd"));
    ASSERT_EQ_INT(-1, sdl_load_mod_texture("C:\\Windows\\system32\\config\\sam"));
    
    // These might bypass current validation
    // ASSERT_EQ_INT(-1, sdl_load_mod_texture("..././etc/passwd"));
    // ASSERT_EQ_INT(-1, sdl_load_mod_texture("....//etc/passwd"));
}
```

File: tests/test_render_primitives.c [hunk 6]
```diff
@@ -417,6 +417,151 @@ TEST(test_stress_many_draws)
...
+
+	// At scale=3
+	sdl_scale = 3;
+	sdl_circle_alpha(100, 100, 50, 0x7FFF, 255, TEST_XOFF, TEST_YOFF);
+
+	// Restore scale
+	sdl_scale = old_scale;
+
+	fprintf(stderr, "     Circle scaling OK\n");
+	ASSERT_TRUE(1);
```
Comment (PrismaPhonic) [inline tests/test_render_primitives.c]: Seems that this tests that things don't crash, but don't actually verify behavior?

**Review by PrismaPhonic — COMMENTED** [review]
Overall this looks really great! I think the main thing is making sure the tests actually test behavior, or if not be explicit about why it's hard to test that behavior. Right now it looks like a few of the tests just verify we don't crash, but don't validate output. I know testing actual sdl output would be pretty hard but we can use clip_line directly to validate behavior. Make it non-static then:

```c
TEST(test_line_clipping_slope)
{
    int x0 = -100, y0 = -100, x1 = 900, y1 = 900;
    int result = clip_line(&x0, &y0, &x1, &y1, 0, 0, 800, 600);
    
    ASSERT_EQ_INT(1, result);
    double slope = (double)(y1 - y0) / (double)(x1 - x0);
    ASSERT_TRUE(slope > 0.99 && slope < 1.01);
}
```

Circle scaling is really hard to test behavior on, so maybe a comment in the test like:

```
    // This test verifies the circle functions don't crash at different scales.
    // Actual correctness verified by code inspection:
    // - sdl_circle_alpha(): sr = radius * sdl_scale
    // - Algorithm uses sr throughout (x = sr, d = 1 - sr)
    // - sdl_circle_filled_alpha(): fsr = radius * sdl_scale
    // - Vertices calculated with fsr
 ```
 
 In summary something like this:
 
 ```c
 TEST(test_circle_scaling)
{
    fprintf(stderr, "  → Testing circle scaling with sdl_scale...\n");
    
    // Actual correctness verified by code inspection:
    // - Circle functions compute sr = radius * sdl_scale
    // - Drawing algorithms use sr throughout
    // This test verifies they don't crash at various scales.
    
    int old_scale = sdl_scale;
    sdl_scale = 1;
    sdl_circle_alpha(100, 100, 50, 0x7FFF, 255, TEST_XOFF, TEST_YOFF);
    sdl_circle_filled_alpha(100, 100, 50, 0x7FFF, 255, TEST_XOFF, TEST_YOFF);
    
    sdl_scale = 2;
    sdl_circle_alpha(100, 100, 50, 0x7FFF, 255, TEST_XOFF, TEST_YOFF);
    sdl_circle_filled_alpha(100, 100, 50, 0x7FFF, 255, TEST_XOFF, TEST_YOFF);
    
    sdl_scale = old_scale;
    fprintf(stderr, "     Circle scaling OK\n");
    ASSERT_TRUE(1);
}

TEST(test_line_clipping_slope)
{
    fprintf(stderr, "  → Testing line clipping preserves slope...\n");
    
    // Test 45-degree diagonal
    int x0 = -100, y0 = -100, x1 = 900, y1 = 900;
    int result = clip_line(&x0, &y0, &x1, &y1, 0, 0, 800, 600);
    ASSERT_EQ_INT(1, result);
    double slope = (double)(y1 - y0) / (double)(x1 - x0);
    ASSERT_TRUE(slope > 0.99 && slope < 1.01);
    
    // Test horizontal line
    x0 = -50; y0 = 300; x1 = 850; y1 = 300;
    result = clip_line(&x0, &y0, &x1, &y1, 0, 0, 800, 600);
    ASSERT_EQ_INT(1, result);
    ASSERT_EQ_INT(300, y0);
    ASSERT_EQ_INT(300, y1);
    
    // Test completely outside
    x0 = -100; y0 = -100; x1 = -50; y1 = -50;
    result = clip_line(&x0, &y0, &x1, &y1, 0, 0, 800, 600);
    ASSERT_EQ_INT(0, result);
    
    fprintf(stderr, "     Line clipping slope preservation OK\n");
}

TEST(test_thick_line_clipping)
{
    fprintf(stderr, "  → Testing thick line clipping...\n");
    
    // Thick lines use clip_line() on center line (tested above)
    // This verifies they don't crash when clipping is needed
    
    sdl_thick_line_alpha(-50, 100, 850, 100, 20, 0x7FFF, 255, 
                         0, 0, 800, 600, TEST_XOFF, TEST_YOFF);
    sdl_thick_line_alpha(-100, -100, -50, -50, 10, 0x7FFF, 255, 
                         0, 0, 800, 600, TEST_XOFF, TEST_YOFF);
    
    fprintf(stderr, "     Thick line clipping OK\n");
    ASSERT_TRUE(1);
}
```
